File: ios/Classes/models/list/FCPListItem.swift
//
//  FCPListItem.swift
//  flutter_carplay
//
//  Created by Oğuzhan Atalay on 21.08.2021.
//

import CarPlay
import Flutter

@available(iOS 14.0, *)
final class FCPListItem {
  private(set) var _super: CPListItem?
  private(set) var elementId: String
  private(set) var text: String?
  private var detailText: String?
  private var isOnPressListenerActive: Bool = false
  private var completeHandler: (() -> Void)?
  private var image: String?
  private var imageData: FlutterStandardTypedData?
  private var imageTint: FCPImageTint?
  private var accessoryImage: String?
  private var trailingImage: String?
  private var trailingImageData: FlutterStandardTypedData?
  private var trailingImageTint: FCPImageTint?
  private var playbackProgress: CGFloat?
  private var isPlaying: Bool?
  private var playingIndicatorLocation: CPListItemPlayingIndicatorLocation?
  private var accessoryType: CPListItemAccessoryType?

  init(obj: [String: Any]) {
    self.elementId = obj["_elementId"] as! String
    self.text = obj["text"] as? String
    self.detailText = obj["detailText"] as? String
    self.isOnPressListenerActive = obj["onPress"] as? Bool ?? false
    self.image = obj["image"] as? String
    self.imageData = obj["imageData"] as? FlutterStandardTypedData
    self.imageTint = FCPImageTint(from: obj["imageTint"] as? [String: Any])
    self.accessoryImage = obj["accessoryImage"] as? String
    self.trailingImage = obj["trailingImage"] as? String
    self.trailingImageData = obj["trailingImageData"] as? FlutterStandardTypedData
    self.trailingImageTint = FCPImageTint(from: obj["trailingImageTint"] as? [String: Any])
    self.playbackProgress = obj["playbackProgress"] as? CGFloat
    self.isPlaying = obj["isPlaying"] as? Bool
    self.setPlayingIndicatorLocation(fromString: obj["playingIndicatorLocation"] as? String)
    self.setAccessoryType(fromString: obj["accessoryType"] as? String)
  }

  private func handler(selectedItem: CPSelectableListItem, complete: @escaping () -> Void) {
    if isOnPressListenerActive {
      completeHandler = complete

      DispatchQueue.main.async {
        FCPStreamHandlerPlugin.sendEvent(
          type: FCPChannelTypes.onListItemSelected,
          data: ["elementId": self.elementId]
        )
      }
    } else {
      complete()
    }
  }

  var get: CPListTemplateItem {
    let listItem = CPListItem.init(text: text, detailText: detailText)
    listItem.handler = self.handler
    if image != nil {
      listItem.setImage(makeSafeUIPlaceholder())
      loadUIImage(from: image!, bytes: imageData, tint: imageTint) { uiImage in
        listItem.setImage(uiImage)
      }
    }

    let accessorySource = trailingImage ?? accessoryImage
    if accessorySource != nil {
      listItem.setAccessoryImage(makeSafeUIPlaceholder())
      loadUIImage(from: accessorySource!, bytes: trailingImageData, tint: trailingImageTint) {
        uiImage in
        listItem.setAccessoryImage(uiImage)
      }
    }

    if playbackProgress != nil {
      listItem.playbackProgress = playbackProgress!
    }
    if isPlaying != nil {
      listItem.isPlaying = isPlaying!
    }
    if playingIndicatorLocation != nil {
      listItem.playingIndicatorLocation = playingIndicatorLocation!
    }
    if accessoryType != nil && accessorySource == nil {
      listItem.accessoryType = accessoryType!
    }
    self._super = listItem
    return listItem
  }

  public func stopHandler() {
    guard self.completeHandler != nil else {
      return
    }
    self.completeHandler!()
    self.completeHandler = nil
  }

  public func update(args: [String: Any]) {
    let text = args["text"] as? String
    let detailText = args["detailText"] as? String
    let image = args["image"] as? String
    let imageData = args["imageData"] as? FlutterStandardTypedData
    let imageTint = FCPImageTint(from: args["imageTint"] as? [String: Any])
    let accessoryImage = args["accessoryImage"] as? String
    let trailingImage = args["trailingImage"] as? String
    let trailingImageData = args["trailingImageData"] as? FlutterStandardTypedData
    let trailingImageTint = FCPImageTint(from: args["trailingImageTint"] as? [String: Any])
    let playbackProgress = args["playbackProgress"] as? CGFloat
    let isPlaying = args["isPlaying"] as? Bool
    let playingIndicatorLocation = args["playingIndicatorLocation"] as? String
    let accessoryType = args["accessoryType"] as? String

    if text != nil {
      self._super?.setText(text!)
      self.text = text!
    }
    if detailText != nil {
      self._super?.setDetailText(detailText)
      self.detailText = detailText
    }

    let imageTintChanged = imageTint != self.imageTint
    if let image = image, image != self.image || imageTintChanged {
      self._super?.setImage(makeSafeUIPlaceholder())
      loadUIImage(from: image, bytes: imageData, tint: imageTint) { uiImage in
        self._super?.setImage(uiImage)
      }
      self.image = image
      self.imageData = imageData
      self.imageTint = imageTint
    } else if image == nil && args.keys.contains("image") {
      self.image = nil
      self.imageData = nil
      self.imageTint = nil
      self._super?.setImage(nil)
    }

    let requestedAccessoryImage = trailingImage ?? accessoryImage
    let currentAccessoryImage = self.trailingImage ?? self.accessoryImage
    let trailingImageTintChanged = trailingImageTint != self.trailingImageTint
    if let requestedAccessoryImage = requestedAccessoryImage,
      requestedAccessoryImage != currentAccessoryImage || trailingImageTintChanged
    {
      self._super?.setAccessoryImage(makeSafeUIPlaceholder())
      loadUIImage(
        from: requestedAccessoryImage,
        bytes: trailingImageData,
        tint: trailingImageTint
      ) { uiImage in
        self._super?.setAccessoryImage(uiImage)
      }
      self.accessoryImage = accessoryImage
      self.trailingImage = trailingImage
      self.trailingImageData = trailingImageData
      self.trailingImageTint = trailingImageTint
    } else if requestedAccessoryImage == nil
      && (args.keys.contains("accessoryImage") || args.keys.contains("trailingImage"))
    {
      self.accessoryImage = nil
      self.trailingImage = nil
      self.trailingImageData = nil
      self.trailingImageTint = nil
      self._super?.setAccessoryImage(nil)
    }

    if playbackProgress != nil {
      self._super?.playbackProgress = playbackProgress!
      self.playbackProgress = playbackProgress
    }
    if isPlaying != nil {
      self._super?.isPlaying = isPlaying!
      self.isPlaying = isPlaying
    }
    if playingIndicatorLocation != nil {
      self.setPlayingIndicatorLocation(fromString: playingIndicatorLocation)
      if self.playingIndicatorLocation != nil {
        self._super?.playingIndicatorLocation = self.playingIndicatorLocation!
      }
    }
    if accessoryType != nil && (self.trailingImage ?? self.accessoryImage) == nil {
      self.setAccessoryType(fromString: accessoryType)
      if self.accessoryType != nil {
        self._super?.accessoryType = self.accessoryType!
      }
    }
  }

  private func setPlayingIndicatorLocation(fromString: String?) {
    if fromString == "leading" {
      self.playingIndicatorLocation = CPListItemPlayingIndicatorLocation.leading
    } else if fromString == "trailing" {
      self.playingIndicatorLocation = CPListItemPlayingIndicatorLocation.trailing
    }
  }

  private func setAccessoryType(fromString: String?) {
    if fromString == "cloud" {
      self.accessoryType = CPListItemAccessoryType.cloud
    } else if fromString == "disclosureIndicator" {
      self.accessoryType = CPListItemAccessoryType.disclosureIndicator
    } else {
      self.accessoryType = CPListItemAccessoryType.none
    }
  }
}

@available(iOS 14.0, *)
extension FCPListItem: FCPListTemplateItem {}

File: android/src/main/kotlin/com/oguzhnatly/flutter_carplay/models/list/FCPListItem.kt
package com.oguzhnatly.flutter_carplay.models.list

import androidx.car.app.model.Action
import androidx.car.app.model.CarIcon
import androidx.car.app.model.ParkedOnlyOnClickListener
import androidx.car.app.model.Row
import com.oguzhnatly.flutter_carplay.Bool
import com.oguzhnatly.flutter_carplay.CGFloat
import com.oguzhnatly.flutter_carplay.CPListItem
import com.oguzhnatly.flutter_carplay.CPListItemAccessoryType
import com.oguzhnatly.flutter_carplay.FCPChannelTypes
import com.oguzhnatly.flutter_carplay.FCPStreamHandlerPlugin
import com.oguzhnatly.flutter_carplay.UIImage
import com.oguzhnatly.flutter_carplay.UIImageObject

/**
 * A wrapper class for CPListItem with additional functionality.
 *
 * @param obj A map containing information about the list item.
 */
class FCPListItem
    (obj: Map<String, Any>) {

    /// The underlying CPListItem instance.
    private lateinit var _super: CPListItem

    /// The unique identifier for the list item.
    var elementId: String
        private set

    /// The primary text of the list item.
    private var text: String

    /// The secondary text of the list item (optional).
    private var detailText: String?

    /// Indicates whether the onPressed event listener is active for the list item.
    private var isOnPressListenerActive: Bool

    /// A closure to be executed when the list item is selected.
    private var completeHandler: (() -> Unit)? = null

    /// The image associated with the list item (optional).
    private var image: UIImage? = null

    /// The playback progress for the list item (optional).
    private var playbackProgress: CGFloat? = null

    /// Indicates whether the list item is in a playing state (optional).
    private var isPlaying: Bool? = null

    /// Indicates whether the list item is enabled (optional).
    private var isEnabled: Bool

//    /// The location of the playing indicator on the list item (optional).
//    private var playingIndicatorLocation: CPListItemPlayingIndicatorLocation?

    /// The accessory type for the list item (optional).
    private var accessoryType: CPListItemAccessoryType? = null

    /// The accessory image associated with the list item (optional).
    private var accessoryImage: UIImage? = null


    init {
        val elementIdValue = obj["_elementId"] as? String
        val textValue = obj["text"] as? String
        assert(elementIdValue != null && textValue != null) {
            "Missing required keys in dictionary for FCPListItem initialization."
        }
        elementId = elementIdValue!!
        text = textValue!!
        detailText = obj["detailText"] as? String
        isOnPressListenerActive = obj["onPressed"] as? Bool ?: false
        playbackProgress = obj["playbackProgress"] as? CGFloat
        isPlaying = obj["isPlaying"] as? Bool ?: false
        isEnabled = obj["isEnabled"] as? Bool ?: true
        image = (obj["image"] as? String)?.let { UIImageObject.fromFlutterAsset(it) }
        accessoryImage =
            (obj["accessoryImage"] as? String)?.let { UIImageObject.fromFlutterAsset(it) }
//        image = UIImage.dynamicImage(
//            lightImage: obj["image"] as? String,
//            darkImage: obj["darkImage"] as? String
//        )
//
//        accessoryImage = UIImage.dynamicImage(
//            lightImage: obj["accessoryImage"] as? String,
//            darkImage: obj["accessoryDarkImage"] as? String
//        )
//
//        setPlayingIndicatorLocation(obj["playingIndicatorLocation"] as? String)
        setAccessoryType(obj["accessoryType"] as? String)
    }


    /** Returns the underlying CPListItem instance configured with the specified properties. */
    fun getTemplate(): CPListItem {

        val onClick = {
            if (isOnPressListenerActive) {
                FCPStreamHandlerPlugin.sendEvent(
                    FCPChannelTypes.onFCPListItemSelected.name, mapOf("elementId" to elementId)
                )
            }
        }

        val builder = Row.Builder().setOnClickListener(ParkedOnlyOnClickListener.create(onClick))
            .setTitle(text).setEnabled(isEnabled)

        detailText?.let { builder.addText(it) }
        image?.let { builder.setImage(it) }
        accessoryImage?.let {
            builder.addAction(
                Action.Builder().setIcon(it).setOnClickListener(onClick).build()
            )
        }

        when (accessoryType) {
            CPListItemAccessoryType.disclosureIndicator -> {
                builder.addAction(
                    Action.Builder().setIcon(CarIcon.BACK).setOnClickListener(onClick).build()
                )
            }

            CPListItemAccessoryType.none, CPListItemAccessoryType.cloud -> {}
            else -> {}
        }

//            playbackProgress?.let { builder.playbackProgress = it }
//            isPlaying?.let { builder.isPlaying = it }
//            playingIndicatorLocation?.let { builder.playingIndicatorLocation = it }

        _super = builder.build()
        return _super
    }


    /** Stops the onPressed event handler for the list item. */
    fun stopHandler() {
        completeHandler?.invoke()
        completeHandler = null
    }

    /**
     * Updates the properties of the list item.
     *
     * @param text The new primary text.
     * @param detailText The new secondary text.
     * @param image The new image.
     * @param darkImage The new dark image.
     * @param accessoryImage The new accessory image.
     * @param accessoryDarkImage The new accessory dark image.
     * @param playbackProgress The new playback progress.
     * @param isPlaying The new playing state.
     * @param playingIndicatorLocation The new playing indicator location.
     * @param accessoryType The new accessory type.
     * @param isEnabled The new enabled state.
     */
    fun update(
        text: String?,
        detailText: String?,
        image: String?,
        darkImage: String?,
        accessoryImage: String?,
        accessoryDarkImage: String?,
        playbackProgress: CGFloat?,
        isPlaying: Bool?,
        playingIndicatorLocation: String?,
        accessoryType: String?,
        isEnabled: Bool?,
    ) {
        text?.let { this.text = it }
        detailText?.let { this.detailText = it }
        isEnabled?.let { this.isEnabled = it }
        accessoryType?.let { setAccessoryType(it) }
        image?.let { this.image = UIImageObject.fromFlutterAsset(it) }
        accessoryImage?.let { this.accessoryImage = UIImageObject.fromFlutterAsset(it) }
//        image?.let { this.image = UIImage.dynamicImage(lightImage= it, darkImage= darkImage) }
//        accessoryImage?.let { this.accessoryImage = UIImage.dynamicImage(lightImage= it, darkImage= accessoryDarkImage) }

//        isPlaying?.let { this.isPlaying = it }
//        playbackProgress?.let { this.playbackProgress = it }
//        playingIndicatorLocation?.let { setPlayingIndicatorLocation(fromString: it) }
    }


    //    private fun setPlayingIndicatorLocation(fromString: String?) {
//        if fromString == "leading" {
//            playingIndicatorLocation = CPListItemPlayingIndicatorLocation.leading
//        } else if fromString == "trailing" {
//            playingIndicatorLocation = CPListItemPlayingIndicatorLocation.trailing
//        }
//    }

    /** Sets the accessory type of the CPListItem based on the provided string value. */
    private fun setAccessoryType(fromString: String?) {
        accessoryType = when (fromString) {
            "none" -> CPListItemAccessoryType.none
            "cloud" -> CPListItemAccessoryType.cloud
            "disclosureIndicator" -> CPListItemAccessoryType.disclosureIndicator
            else -> CPListItemAccessoryType.none
        }
    }
}
